pkg/controllers/management/nodepool/nodepool.go

```diff
@@ -319,12 +320,23 @@ func needRoleUpdate(node *v3.Node, nodePool *v3.NodePool) bool {
 			nodeRolesMap[services.WorkerRole] = true
 		}
 	}
+	if _, ok := nodeRolesMap[services.ETCDRole]; !ok {
+		nodeRolesMap[services.ETCDRole] = false
+	}
+	if _, ok := nodeRolesMap[services.ControlRole]; !ok {
+		nodeRolesMap[services.ControlRole] = false
+	}
+	if _, ok := nodeRolesMap[services.WorkerRole]; !ok {
+		nodeRolesMap[services.WorkerRole] = false
+	}
 
 	poolRolesMap := map[string]bool{}
 	poolRolesMap[services.ETCDRole] = nodePool.Spec.Etcd
 	poolRolesMap[services.ControlRole] = nodePool.Spec.ControlPlane
 	poolRolesMap[services.WorkerRole] = nodePool.Spec.Worker
-	return !reflect.DeepEqual(nodeRolesMap, poolRolesMap)
+
+	r := !reflect.DeepEqual(nodeRolesMap, poolRolesMap)
```

Comment:
might be worth printing the discrepancies in the debug mode. It helped with debugging similar issues in the past

Reply:
What about printing only when it will cause an update?

Reply:
sounds good